piqueserver/core_commands/social.py

```diff
@@ -10,20 +11,21 @@ def login(connection, password):
     """
     if connection not in connection.protocol.players:
         raise KeyError()
-    for user_type, passwords in connection.protocol.passwords.items():
-        if password in passwords:
-            if user_type in connection.user_types:
-                return "You're already logged in as %s" % user_type
-            return connection.on_user_login(user_type, True)
-    if connection.login_retries is None:
-        connection.login_retries = connection.protocol.login_retries - 1
-    else:
-        connection.login_retries -= 1
-    if not connection.login_retries:
-        connection.kick('Ran out of login attempts')
-        return
-    return 'Invalid password - you have %s tries left' % (
-        connection.login_retries)
+    try:
+        user_type = auth.login((connection.name, password))
+        if user_type in connection.user_types:
+            return "You're already logged in as {}".format(user_type)
+        auth.set_user_type(connection, user_type)
+        notify_login(connection, user_type)
+    except AuthError:
+        # kick if out of login attempts
+        log_failed_attempt(connection)
+        left = retries_left(connection)
+        if left < 1:
+            connection.kick('Ran out of login attempts')
+            return
+        # notify with login retries
+        return 'Invalid password - you have {} tries left'.format(left)
```

Comment:
Similarly, I think it would be good for the AuthError to be able to specify a reason for the authentication failure to be shown to the user, in the form `raise AuthError("my reason")`